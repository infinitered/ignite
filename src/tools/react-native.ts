import { GluegunToolbox } from "gluegun"
import { children } from "./filesystem-ext"
import { boolFlag } from "./flag"

export const isAndroidInstalled = (toolbox: GluegunToolbox): boolean => {
  const androidHome = process.env.ANDROID_HOME
  const hasAndroidEnv = !toolbox.strings.isBlank(androidHome)
  const hasAndroid = hasAndroidEnv && toolbox.filesystem.exists(`${androidHome}/tools`) === "dir"

  return Boolean(hasAndroid)
}

type CopyBoilerplateOptions = {
  boilerplatePath: string
  targetPath: string
  excluded: Array<string>
  overwrite?: boolean
}

/**
 * Copies the boilerplate over to the destination folder.
 *
 */
export async function copyBoilerplate(toolbox: GluegunToolbox, options: CopyBoilerplateOptions) {
  const { filesystem } = toolbox
  const { copyAsync, path } = filesystem

  // ensure the destination folder exists
  await filesystem.dirAsync(options.targetPath)

  // rather than copying everything wholesale, let's check what's in the boilerplate folder
  // and copy over everything except stuff like lockfiles and node_modules
  // just to make it faster, y'know? Don't want to copy unnecessary stuff
  const filesAndFolders = children(options.boilerplatePath, true)
  const copyTargets = filesAndFolders.filter(
    (file) => !options.excluded.find((exclusion) => file.includes(exclusion)),
  )

  const { overwrite } = options
  // kick off a bunch of copies
  const copyPromises = copyTargets.map((fileOrFolder) =>
    copyAsync(path(options.boilerplatePath, fileOrFolder), path(options.targetPath, fileOrFolder), {
      ...(overwrite && { overwrite }),
    }),
  )

  // copy them all at once
  return Promise.all(copyPromises)
}

export async function renameReactNativeApp(
  toolbox: GluegunToolbox,
  oldName: string,
  newName: string,
  oldBundleIdentifier: string,
  newBundleIdentifier: string,
) {
  const { parameters, filesystem, print, strings } = toolbox
  const { kebabCase } = strings
  const { path } = filesystem

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  // lower case stuff
  const oldnamelower = oldName.toLowerCase()
  const newnamelower = newName.toLowerCase()

  // kebab case
  const oldnamekebab = kebabCase(oldName)
  const newnamekebab = kebabCase(newName)

  // SCREAMING_SNAKE_CASE
  const oldnamesnake = oldnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()
  const newnamesnake = newnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()

  async function rename(oldFile: string, newFile: string) {
    log(`Renaming ${oldFile} to ${newFile}`)
    return filesystem.renameAsync(oldFile, newFile)
  }

  // rename files and folders

  // prettier-ignore
  await Promise.allSettled([
    rename(`ios/${oldName}.xcodeproj/xcshareddata/xcschemes/${oldName}.xcscheme`, `${newName}.xcscheme`),
    rename(`ios/${oldName}Tests/${oldName}Tests.m`, `${newName}Tests.m`),
    rename(`ios/${oldName}/${oldName}-Bridging-Header.h`, `${newName}-Bridging-Header.h`),
    rename(`ios/${oldName}/${oldName}.entitlements`, `${newName}.entitlements`),
    rename(`ios/${oldName}.xcworkspace`, `${newName}.xcworkspace`),
    rename(`ios/${oldName}`, `${newName}`),
  ])

  // these we delay to avoid race conditions
  await Promise.allSettled([
    rename(`ios/${oldName}Tests`, `${newName}Tests`),
    rename(`ios/${oldName}.xcodeproj`, `${newName}.xcodeproj`),
  ])

  // if the bundle identifier / android package name changed,
  // we need to move everything to the new folder structure
  const oldPath = oldBundleIdentifier.replace(/\./g, "/")
  const newPath = newBundleIdentifier.replace(/\./g, "/")

  if (oldBundleIdentifier !== newBundleIdentifier) {
    log(`Renaming bundle identifier to ${newBundleIdentifier}`)

    // move everything at the old bundle identifier path to the new one
    await Promise.allSettled([
      filesystem.moveAsync(
        `android/app/src/main/java/${oldPath}`,
        `android/app/src/main/java/${newPath}`,
      ),
      filesystem.moveAsync(
        `android/app/src/debug/java/${oldPath}`,
        `android/app/src/debug/java/${newPath}`,
      ),
      filesystem.moveAsync(
        `android/app/src/release/java/${oldPath}`,
        `android/app/src/release/java/${newPath}`,
      ),
    ])
  }

  // here's a list of all the files to patch the name in
  const filesToPatch = [
    `app.json`,
    `package.json`,
    `android/settings.gradle`,
    `android/app/_BUCK`,
    `android/app/BUCK`,
    `android/app/build.gradle`,
    `android/app/src/debug/java/${newPath}/ReactNativeFlipper.java`,
    `android/app/src/release/java/${newPath}/ReactNativeFlipper.java`,
    `android/app/src/main/AndroidManifest.xml`,
    `android/app/src/main/java/${newPath}/MainActivity.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/MainApplicationReactNativeHost.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/components/MainComponentsRegistry.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/modules/MainApplicationTurboModuleManagerDelegate.java`,
    `android/app/src/main/jni/Android.mk`,
    `android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h`,
    `android/app/src/main/jni/MainComponentsRegistry.h`,
    `android/app/src/main/res/values/strings.xml`,
    `ios/Podfile`,
    `ios/main.jsbundle`, // this file could just be regenerated, but this isn't bad to do
    `ios/${newName}/Info.plist`,
    `ios/${newName}.xcodeproj/project.pbxproj`,
    `ios/${newName}.xcodeproj/xcshareddata/xcschemes/${newName}.xcscheme`,
    `ios/${newName}.xcworkspace/contents.xcworkspacedata`,
    `ios/${newName}Tests/${newName}Tests.m`,
    `ios/${newName}/AppDelegate.mm`,
    `ios/${newName}/LaunchScreen.storyboard`,
  ]

  // patch the files
  await Promise.allSettled(
    filesToPatch.map(async (file) => {
      // no need to patch files that don't exist
      const exists = await filesystem.existsAsync(path(file))
      if (!exists) return

      const content = await filesystem.readAsync(path(process.cwd(), file), "utf8")

      log(`Patching ${file} - ${oldName} to ${newName} and variants`)

      // replace all instances of the old name and all its variants
      const newContent = content
        .replace(new RegExp(oldBundleIdentifier, "g"), newBundleIdentifier)
        .replace(new RegExp(oldnamekebab, "g"), newnamekebab)
        .replace(new RegExp(oldnamesnake, "g"), newnamesnake)
        .replace(new RegExp(oldName, "g"), newName)
        .replace(new RegExp(oldnamelower, "g"), newnamelower)

      // write the new content back to the file
      await filesystem.writeAsync(file, newContent, { atomic: true })
    }),
  )
}

export async function replaceMaestroBundleIds(
  toolbox: GluegunToolbox,
  oldBundleIdentifier: string,
  newBundleIdentifier: string,
) {
  const { parameters, filesystem, print } = toolbox
  const { path } = filesystem

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  // here's a list of all the maestro test files to fix the bundle id
  const TARGET_DIR = path(process.cwd())
  const filesToPatch = filesystem.cwd(TARGET_DIR).find({
    matching: `.maestro/**.yaml`,
    files: true,
    directories: false,
  })

  // patch the files
  await Promise.allSettled(
    filesToPatch.map(async (file) => {
      // no need to patch files that don't exist
      const exists = await filesystem.existsAsync(path(file))
      if (!exists) return

      const content = await filesystem.readAsync(path(process.cwd(), file), "utf8")

      log(`Patching ${file} - ${oldBundleIdentifier} to ${newBundleIdentifier} and variants`)

      // replace all instances of the placeholder bundle id with the new one
      const newContent = content.replace(new RegExp(oldBundleIdentifier, "g"), newBundleIdentifier)

      // write the new content back to the file
      await filesystem.writeAsync(file, newContent, { atomic: true })
    }),
  )
}

export function createExpoRouterScreenTemplate(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  try {
    const TARGET_DIR = filesystem.path(process.cwd())
    const filePath = filesystem.path(TARGET_DIR, "ignite/templates/screen/NAME.tsx.ejs")

    const EXPO_ROUTER_SCREEN_TPL = `import React, { FC } from "react"
import { observer } from "mobx-react-lite"
import { ViewStyle } from "react-native"
import { Screen, Text } from "src/components"

// @mst replace-next-line export default function <%= props.pascalCaseName %>Screen() {
export default observer(function <%= props.pascalCaseName %>Screen() {
  return (
    <Screen style={$root} preset="scroll">
      <Text text="<%= props.camelCaseName %>" />
    </Screen>
  )
// @mst replace-next-line }
})

const $root: ViewStyle = {
  flex: 1,
}
`
    filesystem.write(filePath, EXPO_ROUTER_SCREEN_TPL)
  } catch (e) {
    log(`Unable to write screen generator template.`)
  }
}

export function refactorExpoRouterReactotronCmds(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  try {
    const TARGET_DIR = filesystem.path(process.cwd())
    const reactotronConfigPath = filesystem.path(TARGET_DIR, "src/devtools/ReactotronConfig.ts")

    let reactotronConfig = filesystem.read(reactotronConfigPath)
    reactotronConfig = reactotronConfig
      .replace(/import { goBack, resetRoot, navigate }.*/g, 'import { router } from "expo-router"')
      .replace(/navigate\(route as any\).*/g, "router.push(route)")
      .replace(/goBack\(\).*/g, " router.back()")

    // this one gets removed entirely
    const customCommandToRemoveRegex =
      /reactotron\.onCustomCommand\({\s*title: "Reset Navigation State",\s*description: "Resets the navigation state",\s*command: "resetNavigation",\s*handler: \(\) => {\s*Reactotron\.log\("resetting navigation state"\)\s*resetRoot\({ index: 0, routes: \[\] }\)\s*},\s*}\),?\n?/g
    reactotronConfig = reactotronConfig.replace(customCommandToRemoveRegex, "")

    filesystem.write(reactotronConfigPath, reactotronConfig)
  } catch (e) {
    log(`Unable to update ReactotronConfig.`)
  }
}

export function updateExpoRouterSrcDir(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  const TARGET_DIR = filesystem.path(process.cwd())
  const expoRouterFilesToFix = [
    "tsconfig.json",
    // has its own tsconfig, needs updating separately
    "test/i18n.test.ts",
    "test/setup.ts",
    "ignite/templates/model/NAME.ts.ejs",
    "ignite/templates/model/NAME.test.ts.ejs",
    "ignite/templates/component/NAME.tsx.ejs",
  ]
  expoRouterFilesToFix.forEach((file) => {
    const filePath = filesystem.path(TARGET_DIR, file)
    let fileContents = filesystem.read(filePath)
    try {
      fileContents = fileContents.replace(/app\//g, "src/")
      filesystem.write(filePath, fileContents)
    } catch (e) {
      log(`Unable to locate ${file}.`)
    }
  })
}

export function cleanupExpoRouterConversion(toolbox: GluegunToolbox, targetPath: string) {
  const { filesystem } = toolbox

  const workingDir = filesystem.cwd(targetPath)
  workingDir.cwd("src").remove("app.tsx")
  workingDir.move(
    workingDir.path("src", "screens", "ErrorScreen"),
    workingDir.path("src", "components", "ErrorBoundary"),
  )
  workingDir.remove("App.tsx")
  workingDir.remove(workingDir.path("ignite", "templates", "screen", "NAMEScreen.tsx.ejs"))
  workingDir.remove(workingDir.path("ignite", "templates", "navigator"))
  workingDir.remove(workingDir.path("src", "screens"))
  workingDir.remove(workingDir.path("src", "navigators"))
  workingDir.remove("app")
}
