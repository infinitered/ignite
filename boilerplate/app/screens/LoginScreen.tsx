import { observer } from 'mobx-react-lite';
import {
  ComponentType,
  FC,
  useEffect,
  useMemo,
  useRef,
  useState,
  useCallback,
} from 'react';
import { TextInput, TextStyle, ViewStyle } from 'react-native';
import i18n from 'i18next';

import {
  Button,
  Icon,
  Screen,
  Text,
  TextField,
  TextFieldAccessoryProps,
} from '../components';
import { useStores } from '../models';
import { AppStackScreenProps } from '../navigators';

import type { ThemedStyle } from '@/theme';
import { useAppTheme } from '@/utils/useAppTheme';
import { useHeader } from '@/utils/useHeader';

interface LoginScreenProps extends AppStackScreenProps<'Login'> {}

export const LoginScreen: FC<LoginScreenProps> = observer(
  function LoginScreen(_props) {
    const authPasswordInput = useRef<TextInput>(null);

    const [, updateState] = useState({});
    const forceUpdate = useCallback(() => updateState({}), []);

    const [isAuthPasswordHidden, setIsAuthPasswordHidden] = useState(true);
    const [isSubmitted, setIsSubmitted] = useState(false);
    const {
      authenticationStore: {
        authPassword,
        authEmail,
        authFirstName,
        authLastName,
        setAuthFirstName,
        setAuthLastName,
        setAuthEmail,
        setAuthToken,
        setAuthPassword,
        setUserId,
        validationEmailError,
        validationPasswordError,
        validationFirstNameError,
        validationLastNameError,
      },
    } = useStores();

    useHeader(
      {
        leftTx: 'common:lang',
        onLeftPress: () => {
          if (i18n.language.includes('en')) {
            i18n.changeLanguage('zh');
          } else {
            i18n.changeLanguage('en');
          }
          forceUpdate();
        },
      },
      [i18n.language],
    );

    const {
      themed,
      theme: { colors },
    } = useAppTheme();

    useEffect(() => {
      // Here is where you could fetch credentials from keychain or storage
      // and pre-fill the form fields.
      setAuthEmail('');
      setAuthPassword('');

      // Return a "cleanup" function that React will run when the component unmounts
      return () => {
        setAuthPassword('');
        setAuthEmail('');
      };
    }, [setAuthEmail, setAuthPassword]);

    const errorEmail = isSubmitted ? validationEmailError : '';
    const errorPassword = isSubmitted ? validationPasswordError : '';
    const errorFirstName = isSubmitted ? validationFirstNameError : '';
    const errorLastName = isSubmitted ? validationLastNameError : '';

    async function login() {
      setIsSubmitted(true);

      if (validationEmailError) return;

      if (validationPasswordError) return;

      if (validationFirstNameError) return;

      if (validationLastNameError) return;

      setIsSubmitted(false);
      setAuthPassword('');
      setAuthEmail('');

      const token = String(Date.now());
      const userId = '8ee7cbd9-4c47-4aaf-a17b-eb29cc3dcff4'; // uuid.v4();

      setAuthToken(token);
      setUserId(userId);
    }

    const PasswordRightAccessory: ComponentType<TextFieldAccessoryProps> =
      useMemo(
        () =>
          function PasswordRightAccessory(props: TextFieldAccessoryProps) {
            return (
              <Icon
                icon={isAuthPasswordHidden ? 'view' : 'hidden'}
                color={colors.palette.neutral800}
                containerStyle={props.style}
                size={20}
                onPress={() => setIsAuthPasswordHidden(!isAuthPasswordHidden)}
              />
            );
          },
        [isAuthPasswordHidden, colors.palette.neutral800],
      );

    return (
      <Screen
        preset="auto"
        contentContainerStyle={themed($screenContentContainer)}
        safeAreaEdges={['top', 'bottom']}
      >
        <Text
          data-testid="login-heading"
          tx="loginScreen:logIn"
          preset="heading"
          style={themed($logIn)}
        />
        <Text
          data-testid="login-enter-details"
          tx="loginScreen:enterDetails"
          preset="subheading"
          style={themed($enterDetails)}
        />

        <TextField
          testID="login-email"
          value={authEmail}
          onChangeText={setAuthEmail}
          containerStyle={themed($textField)}
          autoCapitalize="none"
          autoComplete="email"
          autoCorrect={false}
          keyboardType="email-address"
          labelTx="loginScreen:emailFieldLabel"
          placeholderTx="loginScreen:emailFieldPlaceholder"
          helper={errorEmail}
          status={errorEmail ? 'error' : undefined}
          onSubmitEditing={() => authPasswordInput.current?.focus()}
        />

        <TextField
          testID="login-first-name"
          value={authFirstName}
          onChangeText={setAuthFirstName}
          containerStyle={themed($textField)}
          autoCapitalize="none"
          autoComplete="name-given"
          autoCorrect={false}
          labelTx="loginScreen:firstNameFieldLabel"
          placeholderTx="loginScreen:firstNameFieldPlaceholder"
          helper={errorFirstName}
          status={errorFirstName ? 'error' : undefined}
        />

        <TextField
          testID="login-last-name"
          value={authLastName}
          onChangeText={setAuthLastName}
          containerStyle={themed($textField)}
          autoCapitalize="none"
          autoComplete="name-family"
          autoCorrect={false}
          labelTx="loginScreen:lastNameFieldLabel"
          placeholderTx="loginScreen:lastNameFieldPlaceholder"
          helper={errorLastName}
          status={errorLastName ? 'error' : undefined}
        />

        <TextField
          testID="login-password"
          ref={authPasswordInput}
          value={authPassword}
          onChangeText={setAuthPassword}
          containerStyle={themed($textField)}
          autoCapitalize="none"
          autoComplete="password"
          autoCorrect={false}
          secureTextEntry={isAuthPasswordHidden}
          labelTx="loginScreen:passwordFieldLabel"
          placeholderTx="loginScreen:passwordFieldPlaceholder"
          onSubmitEditing={login}
          helper={errorPassword}
          status={errorPassword ? 'error' : undefined}
          RightAccessory={PasswordRightAccessory}
        />

        <Button
          testID="login-button"
          tx="loginScreen:tapToLogIn"
          style={themed($tapButton)}
          preset="reversed"
          onPress={login}
        />
      </Screen>
    );
  },
);

const $screenContentContainer: ThemedStyle<ViewStyle> = ({ spacing }) => ({
  paddingVertical: spacing.xxs,
  paddingHorizontal: spacing.lg,
});

const $logIn: ThemedStyle<TextStyle> = ({ spacing }) => ({
  marginBottom: spacing.sm,
});

const $enterDetails: ThemedStyle<TextStyle> = ({ spacing }) => ({
  marginBottom: spacing.lg,
});

const $textField: ThemedStyle<ViewStyle> = ({ spacing }) => ({
  marginBottom: spacing.lg,
});

const $tapButton: ThemedStyle<ViewStyle> = ({ spacing }) => ({
  marginTop: spacing.xs,
});
