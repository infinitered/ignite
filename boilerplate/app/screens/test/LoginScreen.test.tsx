import { screen, render, fireEvent } from '@testing-library/react-native';

import { LoginScreen } from '../LoginScreen';

jest.mock('../../models', () => ({
  useStores: jest.fn(() => ({
    authenticationStore: {
      setUserId: jest.fn(),
      setAuthEmail: jest.fn(),
      setAuthPassword: jest.fn(),
      setAuthToken: jest.fn(),
      validationEmailError: 'error:email',
      validationPasswordError: 'error:password',
      validationFirstNameError: 'error:firstName',
      validationLastNameError: 'error:lastName',
    },
    storageStore: {},
  })),
}));

jest.mock('@/utils/useAppTheme', () => ({
  useAppTheme: jest.fn(() => ({
    themed: () => (Component: any) => Component,
    theme: {
      colors: {
        palette: {
          neutral800: '#000',
        },
      },
    },
  })),
}));

jest.mock('../../utils/useHeader', () => ({
  useHeader: jest.fn(),
}));

describe('LoginScreen', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the LoginScreen - Snapshot', () => {
    render(<LoginScreen navigation={{} as any} route={{} as any} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should display error message if fields are invalid', async () => {
    const { getByTestId, getByText } = render(
      <LoginScreen navigation={{} as any} route={{} as any} />,
    );

    // Press the login button
    const loginButton = getByTestId('login-button');
    fireEvent.press(loginButton);

    expect(getByText('error:email')).toBeTruthy();
    expect(getByText('error:password')).toBeTruthy();
    expect(getByText('error:firstName')).toBeTruthy();
    expect(getByText('error:lastName')).toBeTruthy();
  });
});
