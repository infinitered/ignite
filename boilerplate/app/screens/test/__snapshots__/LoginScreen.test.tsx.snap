// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders the LoginScreen - Snapshot 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "height": "100%",
        "width": "100%",
      },
      {
        "backgroundColor": undefined,
      },
      {
        "paddingBottom": 0,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "flex": 1,
          },
          undefined,
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      bottomOffset={50}
      contentContainerStyle={
        [
          {
            "alignItems": "stretch",
            "justifyContent": "flex-start",
          },
          undefined,
          [Function],
        ]
      }
      keyboardShouldPersistTaps="handled"
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      scrollEnabled={true}
      style={
        [
          {
            "flex": 1,
            "height": "100%",
            "width": "100%",
          },
          undefined,
          undefined,
        ]
      }
    >
      <View>
        <Text
          data-testid="login-heading"
          preset="heading"
          style={
            [
              {},
              [Function],
              undefined,
              undefined,
              [Function],
            ]
          }
        >
          loginScreen:logIn
        </Text>
        <Text
          data-testid="login-enter-details"
          preset="subheading"
          style={
            [
              {},
              [Function],
              undefined,
              undefined,
              [Function],
            ]
          }
        >
          loginScreen:enterDetails
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            preset="formLabel"
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [Function],
              ]
            }
          >
            loginScreen:emailFieldLabel
          </Text>
          <View
            style={[Function]}
          >
            <TextInput
              autoCapitalize="none"
              autoComplete="email"
              autoCorrect={false}
              editable={true}
              keyboardType="email-address"
              onChangeText={
                [MockFunction] {
                  "calls": [
                    [
                      "",
                    ],
                  ],
                  "results": [
                    {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                }
              }
              onSubmitEditing={[Function]}
              placeholder="loginScreen:emailFieldPlaceholder"
              style={[Function]}
              testID="login-email"
              textAlignVertical="top"
            />
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            preset="formLabel"
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [Function],
              ]
            }
          >
            loginScreen:firstNameFieldLabel
          </Text>
          <View
            style={[Function]}
          >
            <TextInput
              autoCapitalize="none"
              autoComplete="name-given"
              autoCorrect={false}
              editable={true}
              placeholder="loginScreen:firstNameFieldPlaceholder"
              style={[Function]}
              testID="login-first-name"
              textAlignVertical="top"
            />
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            preset="formLabel"
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [Function],
              ]
            }
          >
            loginScreen:lastNameFieldLabel
          </Text>
          <View
            style={[Function]}
          >
            <TextInput
              autoCapitalize="none"
              autoComplete="name-family"
              autoCorrect={false}
              editable={true}
              placeholder="loginScreen:lastNameFieldPlaceholder"
              style={[Function]}
              testID="login-last-name"
              textAlignVertical="top"
            />
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            preset="formLabel"
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [Function],
              ]
            }
          >
            loginScreen:passwordFieldLabel
          </Text>
          <View
            style={[Function]}
          >
            <TextInput
              autoCapitalize="none"
              autoComplete="password"
              autoCorrect={false}
              editable={true}
              onChangeText={
                [MockFunction] {
                  "calls": [
                    [
                      "",
                    ],
                  ],
                  "results": [
                    {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                }
              }
              onSubmitEditing={[Function]}
              placeholder="loginScreen:passwordFieldPlaceholder"
              secureTextEntry={true}
              style={[Function]}
              testID="login-password"
              textAlignVertical="top"
            />
            <View
              accessibilityRole="imagebutton"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <RCTImageView
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                internal_analyticTag={null}
                resizeMode="contain"
                source={
                  [
                    {
                      "testUri": "../../../assets/icons/view.png",
                    },
                  ]
                }
                style={
                  [
                    {
                      "height": undefined,
                      "width": undefined,
                    },
                    {
                      "overflow": "hidden",
                    },
                    [
                      {
                        "resizeMode": "contain",
                      },
                      {
                        "tintColor": "#000",
                      },
                      {
                        "height": 20,
                        "width": 20,
                      },
                      undefined,
                    ],
                  ]
                }
                tintColor="#000"
              />
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          preset="reversed"
          style={
            [
              [Function],
              [Function],
              false,
              false,
            ]
          }
          testID="login-button"
        >
          <Text
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [
                  [Function],
                  undefined,
                  false,
                  false,
                ],
              ]
            }
          >
            loginScreen:tapToLogIn
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
